pkg/sql/alter_job_owner.go

```diff
@@ -0,0 +1,168 @@
+		return err
+	}
+	if !v.AtLeast(clusterversion.V25_1.Version()) {
+		return pgerror.Newf(pgcode.FeatureNotSupported, "ALTER JOB OWNER requires version %s", clusterversion.V25_1)
+	}
+
+	exprEval := p.ExprEvaluator("ALTER JOB OWNER")
+	jobIDInt, err := exprEval.Int(ctx, n.jobID)
+	if err != nil {
+		return err
+	}
+	jobID := jobspb.JobID(jobIDInt)
+
+	cur, err := params.ExecCfg().InternalDB.Executor().QueryRowEx(
+		ctx, "get-job-owner", p.txn, sessiondata.NodeUserSessionDataOverride,
+		`SELECT owner FROM system.jobs WHERE id = $1`, jobID,
+	)
+	if err != nil {
+		return err
+	}
+	if cur == nil {
+		return jobs.NewJobNotFoundError(jobID)
+	}
+
+	currentOwner := username.MakeSQLUsernameFromPreNormalizedString(string(tree.MustBeDString(cur[0])))
+
+	if err := n.checkOwnership(ctx, params.p, currentOwner, false); err != nil {
+		return err
+	}
+
+	if err := n.checkOwnership(ctx, params.p, newOwner, true); err != nil {
+		return err
+	}
+
+	if newOwner == currentOwner {
+		return nil
+	}
+
+	if _, err = params.ExecCfg().InternalDB.Executor().ExecEx(
+		ctx, "set-job-owner", p.txn, sessiondata.NodeUserSessionDataOverride,
+		`UPDATE system.jobs SET owner = $2 WHERE id = $1`, jobID, newOwner.Normalized(),
+	); err != nil {
+		return err
+	}
+
+	// Update the legacy payload's copy as well since there are still some paths
+	// which read it.
+	// TODO(dt): remove this when we drop the field from the legacy payload.
+	legacyJob, err := params.ExecCfg().JobRegistry.LoadJobWithTxn(ctx, jobID, p.InternalSQLTxn())
+	if err != nil {
+		return err
+	}
+	return legacyJob.WithTxn(p.InternalSQLTxn()).Update(ctx, func(_ isql.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
+		md.Payload.UsernameProto = newOwner.EncodeProto()
+		ju.UpdatePayload(md.Payload)
+		return nil
+	})
+}
+
+// checkOwnership checks that the current user is allowed to set the owner of a
+// job that is owned or will be owned by the passed owner. This should be called
+// twice: once for the current owner and once for the new owner, with the param
+// forNewOwner indicating which is being done to provide clearer errors.
+func (n *alterJobOwnerNode) checkOwnership(
+	ctx context.Context, p *planner, owner username.SQLUsername, forNewOwner bool,
+) error {
+	// If the user is owner, this is trivially true.
+	if owner == p.User() {
+		return nil
+	}
+
+	// Nobody is allowed to transfer jobs to "node".
+	if owner == username.NodeUserName() {
+		return pgerror.Newf(pgcode.InsufficientPrivilege, "cannot transfer ownership to or from node")
+	}
+
+	// If the user is an admin, ownership is irrelevant.
+	isAdmin, err := p.UserHasAdminRole(ctx, p.User())
+	if err != nil {
+		return err
+	}
+	if isAdmin {
+		return nil
+	}
+
+	// If the user is a member of a role that is the owner, they are an owner.
+	members, err := p.MemberOfWithAdminOption(ctx, p.User())
+	if err != nil {
+		return err
+	}
+	for role := range members {
+		if role == owner {
+			return nil
+		}
+	}
+
+	if forNewOwner {
```

Comment:
nit: instead of passing a `forNewOwner` flag solely for creating an error, you could instead wrap the errors returned by this helper func?

Reply:
ah, you don't want to wrap the pgerror. that's fine.